[RFC 0140] Simple Package Paths (#140)
NixOS/rfcs: rfcs/0140-simple-package-paths.md

---
feature: simple-package-paths
start-date: 2022-09-02
author: Silvan Mosberger

> co-authors: Nixpkgs Architecture Team

Reviewer: A team is not an author, because an author is an individual.
Not just semantics, because we should have a good record of who actually participated in the drafting of this RFC.
Might have to dig through the notes to make this complete...

```suggestion
co-authors: Thomas Bereknyei, Robert Hensing, John Ericson, Alex Ameen
```

Author: Not entirely related to this review, but we discussed how to unblock this RFC in the NAT meeting. Since only me and @roberth contributed text to this RFC, only we are authors, but we can still list the other members of the team as reviewers like this:

```suggestion
co-authors: Robert Hensing
pre-RFC reviewers: Thomas Bereknyei, John Ericson, Alex Ameen
```

This way the reviewers (@aakropotkin said he can do it) can nominate themselves as shepherds without directly violating the RFC rules. I'm not a big fan of doing this, because all the reviewers from the Nixpkgs Architecture Team are obviously already in favor of the RFC. But this RFC has been open for 2 months now and we still don't have enough shepherds (I also reached out [on Discourse](https://discourse.nixos.org/t/shepherds-needed-for-rfc-140/26543)). I don't think this RFC should be blocked by this any longer, especially since a team of 5 people already discussed it for months and reached consensus about it.

Reviewer 2: > I'm not a big fan of doing this, because all the reviewers from the Nixpkgs Architecture Team are obviously already in favor of the RFC.

I think self-nominating, disclosing «currently siding towards FCP/accept after a week or two», and pledging to give serious consideration to any novel arguments for modification/rejection is perfectly fine.

We have had RFCs with long pre-nomination discussions where people self-nominated with a stated position. If you sincerely say that the effective contributions of the remaining team members are comparable to pre-nomination reviews in other RFCs, it's all fine; you did pre-submission round of reviews in a non-standard way, but you should not be burdened too much more than if you had submitted as a draft and then only three people showed up and then you marked it ready for review after a month of polishing. 

I have made a call for «opposing» shepherds, if we get none, maybe it is a sign of something?

Reviewer 3: I don't think there is a meaningful difference between a co-author and someone who was present at the authoring sessions, but did not physically type any of the text. Someone who is part of the NAT will likely not be able to bring a fresh perspective as opposed to someone entirely new.

Author: It feels overly rigorous to me to not be able to progress on a universally-liked idea with 7 reviewers (5 NAT members and 2 shepherds), just because we can't find another person. I'd say we should leave it up to the @NixOS/rfc-steering-committee to decide on this.

Reviewer 4: Our RFC process has the problem that its model of how the process looks like simply does not match reality anymore. I've followed or participated in a couple of RFCs so far, and in many times the proposal was largely reworked or completely rewritten by the shepherd team.

The fact that people organize *before* writing the RFC means that it is likely to be more complete, thought-through and higher quality overall. I see this as a win, and our formalized RFC process actively discourages doing so which is unfortunate.

> Also we measure interest (or lack thereof) by getting together a shepherd team.

I'm sorry, but "interest" is a horrible metric for an RFC. There is a lot of boring, thankless, tedious work to that few people are interested in.

----

I honestly don't know what to say about this. I've been deeply dissatisfied with out current process so far on many accounts, and we should have a larger discussion about this elsewhere. It just would be unfortunate to block RFC 140 on that discussion though.

Reviewer 3: > I'm sorry, but "interest" is a horrible metric for an RFC. There is a lot of boring, thankless, tedious work to that few people are interested in.

https://github.com/NixOS/rfcs/pull/130, what can I say.

Regarding the main point though, to play devil's advocate: Say you have 5 people in the NAT, say you have an RFC that has an author from the NAT (nominally or actually) — then you also have 4 shepherds and a conveniently regularly scheduled meeting of the shepherds — you can go straight to FCP basically.

Reviewer 4: Honestly? I don't have a problem with that. IME people don't care and give feedback until FCP anyways …

Author: Thanks a lot for [your nomination](https://github.com/NixOS/rfcs/pull/140#issuecomment-1512080692) as a shepherd @piegamesde! This avoids this problem for now, but we should really have a more proper discussion about this in the future. I'll consider creating an RFC to improve the RFC process around this.

shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: https://github.com/NixOS/nixpkgs/pull/211832
---

# Summary
[summary]: #summary

Auto-generate trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix` from a directory structure that matches the attribute name.

> This makes it much easier to contribute new packages packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.

Reviewer 5: ```suggestion
This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.
```

# Motivation
[motivation]: #motivation

> - (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:

Reviewer 3: Playing devil's advocate: Contributing *new* packages is the last thing that should be made easier for first time contributors, since they are less likely to stick around and actually maintain their additions.

Reviewer 6: ```suggestion
- Contributors of new packages, especially Nix and Nixpkgs beginners, are having a hard time figuring out which files to add or edit. These are very common questions:
```

Reviewer 2: As a rude elitist, I note that addressing some usability issues but not the others allows us to influence what type of contributions will happen as a part of the learning process and which contributions usually come from experienced collaborators.

On the other hand, there is «you need to spend effort learning this, then it starts making sense», and there is «you will learn that this is not actually consistent so you can just guess». I believe here we have the latter case…

```suggestion
- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
+It is not obvious to package contributors where to add files. Even experienced contributors don't make the choices in the same way. These are very common questions:
```

Author: I don't think we need to be very nit-picky with the sentencing here. Just to avoid the ambiguity I'll change it to:

> It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:

@sternenseemann This change doesn't address your original comment though, but with the discussion here, do you consider that addressed?

  - Which directory should my package definition go in?
  - What are all the categories and do they matter?
  - What if the package has multiple matching categories?
  - Why can't I build my package after adding the package file?
  - Where in all-packages.nix should my package go?
- Figuring out where an attribute is defined is a bit tricky:
  - First one has to find the definition of it in all-packages.nix to see what file it refers to
    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
    - It also slows down or even deadlocks editors due to the file size
  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

> This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs, where `shard = toLower (substring 0 2 name)`.

Reviewer 3: What's that name about?

Author: The `name` is the same as the one in `pkgs.${name}`, it's the attribute name

Reviewer 3: I mean `unit`.

Author: See https://github.com/nixpkgs-architecture/rfcs/blob/master/rfcs/0140-simple-package-paths.md#alternate-pkgsunit-location

All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.

The following requirements will be checked by CI.
This standard must be followed for newly added packages that can satisfy these requirements.
A treewide migration to this standard will be performed for existing packages that can satisfy these requirements.

## Structure

The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.

This ensures that maintainers don't have to verify this structure manually, which is prone to mistakes.

## Only derivations

If `pkgs/unit/${shard}/${name}` exists, `pkgs.${name}` must be a derivation that can be built directly with `nix-build`.

This ensures that people can expect the unit directories to correspond to buildable packages and not functions like `pkgs.fetchFromGitHub` or `pkgs.buildRustCrate`.

> ## Stable boundary

Reviewer 3: How do you plan to enforce this? No ofborg check, no boundary.

Author: Tooling to enforce this is being worked on [here](https://github.com/nixpkgs-architecture/nix-spp).

Roughly it works by parsing all Nix files in nixpkgs, figuring out which other files they (statically) reference, and erroring when those references cross unit boundaries. This will be the same code that performed the auto-migration in [the draft implementation](https://github.com/NixOS/nixpkgs/pull/211832), which notably skipped packages that couldn't be safely moved without breaking references.

It's not 100% perfect because files can also be referenced dynamically, e.g. this (made-up) function `subpath: import (dirOf ./. + "subpath")`, but within nixpkgs static references (like `import ./some/file.nix`) are most common, so it works out.

Unit directories may only interact with the rest of Nixpkgs via the stable `pkgs.${name}` attributes, not with file references:

> - Files inside a unit directory must not reference files outside that unit directory.

Reviewer 7: how does it work with update scripts that do e.g. `-I nixpkgs=../../../.`

Author: Yeah that's a good point, it was also mentioned in the Matrix channel. I should add a generic mention that such references should be ensured to not break. `CODEOWNERS` is another such case. This should be doable semi-manually, since it's not done that often (same for such update scripts).

Reviewer 4: I'd say we should rather define a standard for how update script are called, which defines in which directory the script will be run and –if necessary– how to find the current nixpgks' path. Maybe write two sentences about this into the "Future work" section?

-----

For CODEOWNERS, I think this should be noted in the drawbacks section, because people/teams won't be able to subscribe to specific package ecosystems like Python or Gnome anymore.

Reviewer 8: Replace with calls to

```
git rev-parse --show-toplevel

Author: @mweinelt: Doesn't work in shebangs like [here](https://github.com/NixOS/nixpkgs/blob/cb2d5a2fa9f2fa6dd2a619fc3be3e2de21a6a2f4/pkgs/applications/version-management/cz-cli/generate-dependencies.sh). I added a mention of this to the RFC though, will definitely have to make sure these don't break. I don't think it's in scope for this RFC to improve the update script situation, I'll add it to future work though.

## Custom arguments

If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be a freely chosen expression.

This ensures that even if a package initially doesn't require a custom `args`, if it later does, it doesn't have to be moved out of the `pkgs/unit` directory to pass custom arguments.

## Examples
[examples]: #examples

To add a new package `pkgs.foobar` to nixpkgs, one only needs to create the file `pkgs/unit/fo/foobar/pkg-fun.nix`.
No need to find an appropriate category nor to modify `pkgs/top-level/all-packages.nix` anymore.

With many packages, the `pkgs/unit` directory may look like this:

```
pkgs
└── unit
   ├── _0
   │  ├── _0verkill
   │  └── _0x
   ┊
   ├── ch
   │  ├── ChowPhaser
   │  ├── CHOWTapeModel
   │  ├── chroma
   │  ┊
   ┊
   ├── t
   │  └── t
   ┊
```

# Interactions
[interactions]: #interactions

> - `nix edit` and search.nixos.org are unaffected, since they rely on `meta.position` to get the file to edit, which still works

Reviewer 3: We should even improve it and calculate `meta.position` without relying on unsafe builtins.

Author: Agreed, we think it's best for this to not be part of the RFC though to keep the scope small. Once this RFC is implemented, changing `meta.position` to use the unit directories might not even need an RFC since it shouldn't be controversial.

> - `git blame` locally and on GitHub is unaffected, since it follows file renames properly.

Author: ```suggestion
- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
- Existing PR's that change migrated files won't have to be updated for the same reason
- For the same reason, changes to the migrated files can be backported to the previous release without problems
```

- A commonly recommended way of building package directories in nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
  Since the path changes `pkg-fun.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/pkg-fun.nix {}'`, which is harder for users.
  However, calling a path like this is an anti-pattern anyway, because it doesn't use the correct nixpkgs version and it doesn't use the correct argument overrides.
  The correct way of doing it was to add the package to `pkgs/top-level/all-packages.nix`, then calling `nix-build -A hello`.
  This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
  By teaching users that `pkgs/unit/*/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.

> # Drawbacks

Reviewer 3: It is no longer possible to pass custom arguments with `callPackage` directly. This is done often enough in `all-packages.nix` and often with good reason.

I know this is mentioned in the Alternatives section, but I'd say it's a genuine drawback. I think such a change needs to have the ambition of being a global solution for everything it applies to (i.e. direct derivation children of `pkgs`, leaving out stuff related to toolchain logic etc.) which it is not in this form.

Author: That is still possible, see the [Custom Arguments](https://github.com/nixpkgs-architecture/rfcs/blob/269ac4f669302c9b09075a40632b664158908a0b/rfcs/0140-simple-package-paths.md#custom-arguments) section. Is this what you mean?

Author: I'm also adding the drawback that it only applies to top-level packages, and only ones defined with `callPackage`.

[drawbacks]: #drawbacks

- The existing categorization of packages gets lost. Counter-arguments:
  - It was never that useful to begin with
    - The categorization was always incomplete, because packages defined in the language package sets often don't get their own categorized file path.
    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
    - Many packages fit multiple categories, leading to multiple locations to search through instead of one

>   - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)

Reviewer 9: It is a bit of circular, because it relies on the other distros doing that categorization. Indeed as far as I remember the BSD Ports family does precisely what we do - use the filesystem layout as metadata.

On the other hand I concede that using the filesystem as metadata is not a good idea for Nixpkgs. I want to propose a solution for this - not here, obviously, but in another RFC.

So, I suggest to include this as "future work".

Author: Nice RFC! I'll mention it in future work :D

>   - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)

Reviewer 10: How would you do that?

For example in https://github.com/NixOS/nixpkgs/tree/master/pkgs/games i see nix packages related to "games". How would i see such a list on Repology?

I think having a category as metadata for a package would be a good future improvement. We can for example extract it from `menu-spec` of Freedesktop.org Specifications (https://specifications.freedesktop.org/menu-spec/latest/apa.html) and show it in https://search.nixos.org/ and use for `nix search`.

Not sure if the current categories have any value and should be preserved. But we have them in git history, so they could get extracted after a change.

Reviewer 6: As you say, these categories only have value in certain contexts. The Nix language is much more flexible for providing that functionality, and all of that can easily move into `meta`.

  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.

## Alternate `pkg-fun.nix` filename

- `default.nix`:
  - Bad because it doesn't have its main benefits here:
    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/unit/hell/hello` equally broken regardless of file name.
  - Not using `default.nix` frees up `default.nix` for a short expression that is actually buildable, e.g. `(import ../.. {}).hello`, although we don't yet have a use case for this that isn't covered by `nix-build ../.. -A $attrname`.
  - Bad because using `default.nix` would tempt users to invoke `nix-build .` whereas making package functions auto-callable is a known anti-pattern as it duplicates the defaults.
  - Good because `default.nix` is already a convention most people are used to

> - `package.nix`/`pkg.nix`: Bad, because it makes the migration to a non-function form of overridable packages harder in the future.

Author: No migration is planned, in fact I don't think this idea of a non-function form has been discussed at all. So I think these names should be reconsidered. Note the discussion around naming in https://github.com/NixOS/rfcs/pull/140#discussion_r1090915765 as well. It will be discussed in the next [NAT meeting](https://github.com/nixpkgs-architecture/#meetings) on monday.

We do plan to make improvements to package declarations in general though, especially for specifying dependencies, which will require changes to the "package function"-style declarations. I think what happens depends on whether this unit directory standard is standardized and opened to third-parties, or whether it's still only internal to Nixpkgs (which is what this RFC describes). If it's standardized we can't really make changes to the structure and semantics of these files anymore. We can still do that if it's Nixpkgs-internal though.

## Alternate `pkgs/unit` location

- Use `unit` (at the Nixpkgs root) instead of `pkgs/unit`.
  This is future proof in case we want to make the directory structure more general purpose, but this is out of scope

> - Other name proposals were deemed worse: `pkg`, `component`, `part`, `mod`, `comp`

Reviewer 7: I would like to nominate `auto`, to me this is more future-proof than `unit` so we can include multiple packages in one directory

Author: > so we can include multiple packages in one directory

How do you mean that?

Reviewer 7: For example, we currently have `element-desktop` and `element-web` that share the same `pin.json` (or `generic.nix` in some other cases), in the future we might want to allow that to be defined in `pkgs/unit`, but I'm not quite sure the name `unit` fits the behavior of having multiple packages under the same directory, and `auto` is more future-proof in that case IMO

Author: I see, yeah so this has been discussed before. I believe in the team the consensus was that multiple versions and variants should eventually end up in the same `pkgs.element` attribute, once we have the design for that figured out. In that case, all those variants would naturally be together in the `pkgs/unit/el/element` directory.

That's also kind of why `unit` was chosen, we talked about [explaining it better](https://github.com/nixpkgs-architecture/simple-package-paths/issues/16) before, something like:

> unit: A collection of standardized files related to the same software component

Reviewer 7: that makes sense to me, thanks for the explanation

Reviewer 4: I don't dislike "auto" but I'd like to note that the name implies the package is *automatically* called. This means that packages with custom arguments would not fit in there anymore, unlike with the current proposal. Maybe they'd be put into "pkgs/manual" instead then? (which then opens up to confusion with manual as in documentation, etc.)

> - While this RFC doesn't address expressions where the second `callPackage` argument isn't `{}`, there is an easy way to transition to an argument of `{}`: For every attribute of the form `name = attrs.value;` in the argument, make sure `attrs` is in the arguments of the file, then add `name ? attrs.value` to the arguments. Then the expression in `all-packages.nix` can too be auto-called
>   - Don't do this for `name = value` pairs though, that's an alias-like thing

`callPackage` does not favor the default argument when both a default argument and a value in `pkgs` exist. Changing the semantics of `callPackage` is out of scope.

## Allow `callPackage` arguments to be specified in `<unit>/args.nix`

The idea was to expand the auto-calling logic according to:

Unit directories are automatically discovered and transformed to a definition of the form
```
# If args.nix doesn't exist
pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix {}
# If args.nix does exists
pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix (import ${unitDir}/args.nix pkgs);
```

Pro:
 - It makes another class of packages uniform, by picking a solution with restricted expressive power.

Con:

>  - It does not solve the contributor experience problem of having to many rules.

Reviewer 3: ```suggestion
 - It does not solve the contributor experience problem of having too many rules.
```

 - `args.nix` is another pattern that contributors need to learn how to use, as we have seen that it is not immediately obvious to everyone how it works.
 - A CI check can mitigate the possible lack of uniformity, and we see a simple implementation strategy for it.
 - This keeps the contents of the unit directories simple and a bit more uniform than with optional `args.nix` files.

# Unresolved questions

> [unresolved]: #unresolved-questions

Reviewer 3: One interesting thing is that for everyone of those “unit” packages can know their attribute name (even statically). This is interesting for evaluation errors where the attribute name can often help you over just the derivation name.

Author: Yeah, though it could be argued that unit directories knowing their own name is an anti-pattern, because moving unit directories around shouldn't influence their contents. This is similar to how it's problematic to do `src = ./.`, because the directory name will be used as the derivation name, therefore changing the result when you just `mv` the repository around (without even changing the contents).

I don't think this needs to be pointed out in the RFC text though

> # Future work

Author: On todays call @gytis-ivaskevicius mentioned, and other people mentioned this in the past as well, that an API endpoint to get the functions directly, without auto-calling them, would be useful. This should be considered future work, @gytis-ivaskevicius calls them "blueprints", see also https://github.com/NixOS/nixpkgs/pull/208020#issuecomment-1419168718

Author: I'm adding this to future work in the next commit

  - Where in all-packages.nix should my package go?
- Figuring out where an attribute is defined is a bit tricky:
  - First one has to find the definition of it in all-packages.nix to see what file it refers to
    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
    - It also slows down or even locks up editors due to the file size
  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in Nixpkgs, where `shard = toLower (substring 0 2 name)`.

> All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.

Author: @piegamesde See here, that's exactly what's done. `pkg-fun.nix` is just the old `default.nix` (we could also call it `default.nix`, see alternatives section). And note how the second argument is `{}`, meaning the auto-calling only works for packages that don't require any arguments. However see the `Custom arguments` section, which allows you to also call the unit directories from `all-packages.nix` if you need to pass a non-`{}` argument.

  - Bad because it makes `git` and file listings slower
- Use `substring 0 3 name` or `substring 0 4 name`. This was not done because it still leads to directories in `pkgs/unit` containing more than 1 000 entries, leading to the same problems.
- Use multi-level structure, like a 2-level 2-prefix structure where `hello` is in `pkgs/unit/he/ll/hello`,
  if packages are less than 4 characters long, we will it out with `-`, e.g. `z` is in `pkgs/unit/z-/--/z`.
  This is not great because it's more complicated, longer to type and it would improve performance only marginally.
- Use a dynamic structure where directories are rebalanced when they have too many entries.
  E.g. `pkgs.foobar` could be in `pkgs/unit/f/foobar` initially.
  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/unit/fo/foobar`.
  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.

> ## Alternate `package.nix` filename

Reviewer 2: Should `pkg-fun.nix` now get a mention as a considered alternative?

Author: Done

## Allow `callPackage` arguments to be specified in `<unit>/args.nix`

The idea was to expand the auto-calling logic according to:

Unit directories are automatically discovered and transformed to a definition of the form
```
# If args.nix doesn't exist
pkgs.${name} = pkgs.callPackage ${unitDir}/package.nix {}
# If args.nix does exists
pkgs.${name} = pkgs.callPackage ${unitDir}/package.nix (import ${unitDir}/args.nix pkgs);
```

Pro:
 - It makes another class of packages uniform, by picking a solution with restricted expressive power.

Con:
 - It does not solve the contributor experience problem of having too many rules.
 - `args.nix` is another pattern that contributors need to learn how to use, as we have seen that it is not immediately obvious to everyone how it works.
 - A CI check can mitigate the possible lack of uniformity, and we see a simple implementation strategy for it.
 - This keeps the contents of the unit directories simple and a bit more uniform than with optional `args.nix` files.

# Unresolved questions
[unresolved]: #unresolved-questions

# Future work

> [future]: #future-work

Reviewer 2: In the discussion above, adding tagging metadata was proposed as possible future work.

Author: Added

This makes it much easier to contribute new packages packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.

# Motivation
[motivation]: #motivation

- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
  - Which directory should my package definition go in?
  - What are all the categories and do they matter?
  - What if the package has multiple matching categories?
  - Why can't I build my package after adding the package file?
  - Where in all-packages.nix should my package go?
- Figuring out where an attribute is defined is a bit tricky:
  - First one has to find the definition of it in all-packages.nix to see what file it refers to
    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)

>     - It also slows down or even deadlocks editors due to the file size

Author: Changed in https://github.com/NixOS/rfcs/pull/140/commits/21c3493a2b64e5987515d45e5590ee5c6429348c

  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs, where `shard = toLower (substring 0 2 name)`.
All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.

The following requirements will be checked by CI.
This standard must be followed for newly added packages that can satisfy these requirements.
A treewide migration to this standard will be performed for existing packages that can satisfy these requirements.

## Structure

The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.

> Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.

Reviewer 11: IMO stick to ``default.nix``.

Reviewer 12: Do note the discussion in https://github.com/nixpkgs-architecture/rfcs/blob/master/rfcs/0140-simple-package-paths.md#alternate-pkg-funnix-filename

Reviewer 10: I don't like `pkg-fun.nix`. It's complicated because it's 2 words that are abbreviated. My first interpretation was "fun with packages" but i guess it stands for "function".

I think one englisch word would be better as file name. "package" is obvious.

I don't understand the explanation why that name is a bad idea.

Reviewer 6: I agree with @davidak on the potential for confusion, and for something that will be so ubiquitous we should take the greatest care not to throw people off. But I can follow the argument to leave `pkg` or `package` for future extensions, and I also think one should be precise with the name to say what it means. And right now, packages are functions.

I suggest either a minimal change that makes it unambiguously "it's not fun":

```suggestion
Each unit directory must contain at least a `pkg-func.nix` file, but may contain arbitrary other files and directories.
```

or, and I find that idea better but it would be on you to specify it further, give the mechanism a name and call the file by that name, such as `callPackage` for the convention how the thing is used:

```suggestion
Each unit directory must contain at least a `callPackage.nix` file, but may contain arbitrary other files and directories.
```

Reviewer 2: I like `callPackage.nix` more than `pkg.*-f.*n.*` variations

Author: I do like the suggestion of `pkg-func.nix`, but I can also see how a single word might be even better, not sure about `callPackage.nix`. I'll bring this up in the next [NAT meeting](https://github.com/nixpkgs-architecture/#meetings) on Monday.

Reviewer 12: `pkg-func` is with with me

Reviewer 9: `pkg-function.nix`, anyone?

Reviewer 7: I also prefer `pkg-function.nix` over `pkg-fun.nix`, IMO `function` shouldn't be abbreviated as it can be hard coming to an agreement trying to choose from `fn//fun/func`.

I would also suggest `lambda.nix` and `pkg-lambda.nix` as alternatives because lambda is the type of the expression

```console
nix-repl> builtins.typeOf ({}: 42)
"lambda"
```

Author: Note also the discussion with @davidak [here](https://github.com/NixOS/rfcs/pull/140#discussion_r1098667630). I would like to reconsider `package.nix`.

I would also like to point out that these unit directories will arguably need versioning at some point (especially once they become a stable API for third-parties (not in scope for this RFC)). This will likely involve a `version.nix` file, where we could declare the version of the unit directory. With version 2, the `package.nix` file could have a different structure, deprecating the old one. Because of that I don't think we need to be very worried about using `package.nix`, so this is my current favorite.

Author: We just had the NAT meeting and reached consensus that `package.nix` is good, [notes here](https://discourse.nixos.org/t/nixpkgs-architecture-team-meeting-29/25439#pkg-funnix-vs-other-nameshttpsgithubcomnixosrfcspull140discussion_r1090915765-2) :slightly_smiling_face:

Author: I just pushed [a commit](https://github.com/NixOS/rfcs/pull/140/commits/967beb129b463a6096f1d3c74bc3d37343439904) to rename this to `package.nix`, I'll resolve this thread

  This ensures that files within unit directories (except `pkg-fun.nix`) can be freely moved and changed without breaking any other packages.

The only notable exception to this rule is the `pkgs/top-level/all-packages.nix` file which may reference the `pkg-fun.nix` file according to the next requirement.

## Custom arguments

If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be a freely chosen expression.

This ensures that even if a package initially doesn't require a custom `args`, if it later does, it doesn't have to be moved out of the `pkgs/unit` directory to pass custom arguments.

## Examples
[examples]: #examples

To add a new package `pkgs.foobar` to nixpkgs, one only needs to create the file `pkgs/unit/fo/foobar/pkg-fun.nix`.
No need to find an appropriate category nor to modify `pkgs/top-level/all-packages.nix` anymore.

With many packages, the `pkgs/unit` directory may look like this:

```
pkgs
└── unit
   ├── _0
   │  ├── _0verkill

>    │  └── _0x

Reviewer 11: Probably a bit late for this discussion but why do we carry the underscore into the filenames? We only need it for the variable name AFAIK.

Author: I'm not sure what you mean, the package is `pkgs._0verkill`, that's why the directory is `_0verkill`, which is why the parent directory is `_0`. Which alternative are you suggesting?

Reviewer 11: We only really need it for the attr but it might be a good idea to be consistent with the directory name.

Yeah, we should keep it like this.

- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)

# Alternatives
[alternatives]: #alternatives

## Alternate `pkgs/unit` structure

- Use a flat directory, e.g. `pkgs.hello` would be in `pkgs/unit/hello`.
  - Good because it's simpler, both for the user and for the code
  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
    - With an optimized `readDir` this isn't much of a problem
  - Bad because the GitHub web interface only renders the first 1'000 entries (and we have about 10'000 that benefit from this transition, even given the restrictions)
  - Bad because it makes `git` slower ([TODO: By how much?](https://github.com/nixpkgs-architecture/simple-package-paths/issues/18))

>   - Bad because directory listing slows down with many files

Reviewer 11: Those are probably based on the same root causes and it is really hard to quantify the numbers because they depend on hardware, filesystem, caching and other running programs.

Author: https://github.com/NixOS/rfcs/commit/269ac4f669302c9b09075a40632b664158908a0b#diff-d795c40df63067333bb77ad1be78f1ea7e46350e124528338d970662a37f6b3bL136-R134

- Use `substring 0 3 name` or `substring 0 4 name`. This was not done because it still leads to directories in `pkgs/unit` containing more than 1'000 entries, leading to the same problems.
- Use multi-level structure, like a 2-level 2-prefix structure where `hello` is in `pkgs/unit/he/ll/hello`,
  if packages are less than 4 characters long, we will it out with `-`, e.g. `z` is in `pkgs/unit/z-/--/z`.

>   This is not great because it's more complicated and it would improve git performance only marginally.

Reviewer 11: also it would be rather annoying to type

Author: https://github.com/NixOS/rfcs/commit/269ac4f669302c9b09075a40632b664158908a0b#diff-d795c40df63067333bb77ad1be78f1ea7e46350e124528338d970662a37f6b3bR138

- Use a dynamic structure where directories are rebalanced when they have too many entries.
  E.g. `pkgs.foobar` could be in `pkgs/unit/f/foobar` initially.
  But when there's more than 1'000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/unit/fo/foobar`.
  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.

## Alternate `pkg-fun.nix` filename

- `default.nix`: Bad because:
  - Doesn't have its main benefits in this case:
    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/unit/hell/hello` equally broken regardless of file name.

>   - Not using `default.nix` frees up `default.nix` for a short expression that is actually buildable, e.g. `(import ../..).hello`, although at that point it might better be auto-generated or implicit in the CLI

Reviewer 11: good because it is consistent with everything so far, we don't need to explain to ~~new~~ contributors why those files are named different, muscle memory.

Author: I added the fact that it's an existing convention as an argument in https://github.com/NixOS/rfcs/commit/269ac4f669302c9b09075a40632b664158908a0b#diff-d795c40df63067333bb77ad1be78f1ea7e46350e124528338d970662a37f6b3bR152

>   - Don't do this for `name = value` pairs though, that's an alias-like thing

`callPackage` does not favor the default argument when both a default argument and a value in `pkgs` exist. Changing the semantics of `callPackage` is out of scope.

# Allow `callPackage` arguments to be specified in `<unit>/args.nix`

The idea was to expand the auto-calling logic according to:

Unit directories are automatically discovered and transformed to a definition of the form
```
# If args.nix doesn't exist
pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix {}
# If args.nix does exists
pkgs.${name} = pkgs.callPackage ${unitDir}/pkg-fun.nix (import ${unitDir}/args.nix pkgs);
```

Pro:
 - It makes another class of packages uniform, by picking a solution with restricted expressive power.

Con:
 - It does not solve the contributor experience problem of having to many rules.
 - `args.nix` is another pattern that contributors need to learn how to use, as we have seen that it is not immediately obvious to everyone how it works.
 - A CI check can mitigate the possible lack of uniformity, and we see a simple implementation strategy for it.

>  - This keeps the contents of the unit directories simple and a bit more uniform than with optional `args.nix` files.

Reviewer 11: Does this solve that the arguments are not well typed and not machine discoverable? If we go this route, we should try to address this.

Author: No it wouldn't fix that problem. This is definitely something for the Nixpkgs Architecture Team though. Definitely out of scope for this RFC though.

Reviewer 2: Types without Nix-side support would be a large performance hit, no? 

I wonder how much discoverability of arguments we can get even with the current structure…

  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.

## PR 1: The unit directory standard

This part establishes the new _unit directory standard_ in Nixpkgs.
This standard is internal to Nixpkgs and not exposed as public interface.
This standard must be documented in the Nixpkgs manual.

### File structure

Create the initially-empty directory `pkgs/unit`, called _unit base directory_, in Nixpkgs.

Check the following using CI:
- The unit base directory must only contain subdirectories of the form `pkgs/unit/${shard}/${name}`, called _unit directories_.
- `name` is a string only consisting of the ASCII characters a-z, A-Z, 0-9, `-` or `_`.
- `shard` is the lowercased first two letters of `name`, expressed in Nix: `shard = toLower (substring 0 2 name)`.

> - Each unit directory must contain a `package.nix` file and may contain arbitrary other files.

Author: A proposal relating to the changes [to remove custom arguments]() and [not enforcing the standard]():

Both the current directory structure and `all-packages.nix` suck, but as @piegamesde also already hinted at, we could just have the unit directory structure without the extra semantic checks, only having these highlighted checks.

While the extra semantics are useful and there for a reason, a lot of attributes don't satisfy them (non-packages, package sets, custom `callPackage` arguments, other `callPackage`s, etc.). According to the current RFC this would not allow all of those attributes to be moved to `pkgs/unit`.

So how about having a _separate_ directory, call it something like `pkgs/sem-less-unit` or `pkgs/manual-unit`. This directory is the same as `pkgs/unit`, but it doesn't get auto-called, you still need to manually edit `all-packages.nix`, but in turn it doesn't come with the restrictions of `pkgs/unit`, allowing all top-level attributes to be migrated to it.

The one semantic restriction we should still want to have though is the one preventing unit directories from accessing files outside themselves, because afaik all such problems are fixable in a better way.

Reviewer: > So how about having a _separate_ directory, call it something like `pkgs/sem-less-unit` or `pkgs/manual-unit`. This directory is the same as `pkgs/unit`,

That's over-engineered and incompatible with our goal of having a single place to put a package.

### Semantics

Introduce code to automatically define `pkgs.${name}` for each unit directory as a value equivalent to
```nix

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Reviewer 13: Glad to see the use of non-empty arguments to `callPackage` going away.

> users viewing a unit directory can be sure that ... the functions arguments correspond directly to attributes in `pkgs`, no exceptions

Don't the `xyzSupport`/`withXyz` booleans conflict with this?

I think "can be sure that functions arguments *without a default value* correspond directly to attributes in `pkgs`" is true though.

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Author: Oh good point, I fully agree.

It gets a bit more interesting though, because defaults get overridden when packages of the same name do exist. Relevant is [this PR of mine](https://github.com/NixOS/nixpkgs/pull/131271) which would enforce attributes with defaults to not exist (so the default applies).

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Reviewer: > I also removed the exception

No, you are introducing an exception, by disallowing something.

Custom arguments are needed _at times_, and if something causes you to need one, you will have no option but to move your package out of `unit` and into the legacy hierarchy. When the problem has been resolved, you could move it back, but why even bother. `unit` has failed you for reasons outside your control.

> It would also be weird just to allow this one exception of `callPackage` with custom arguments, when we could also make an exception for custom `callPackage`'s `python3Packages.callPackage`, `haskellPackages.callPackage`, or generally any expression referencing the file.

It is not because the need for those exceptions doesn't vary over time; only by package.

We can easily tighten up the rules when we actually have the opportunity to do so - when we have a good alternative way to provide default arguments, but until then this restriction is counterproductive.

> And the user doesn't have to look at `all-packages.nix`, which is the important part.

If this is a goal, we should not have packages as functions in the first place. Without a better default arguments store, this is not a realistic goal.

Improving the default arguments in package functions is out of scope for this RFC. No _package behavior_ should be affected by this RFC.

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Reviewer 4: I agree that requiring no custom arguments is not feasible right now, because it would require moving the package in and out of the folder when the arguments change, which occasionally happens.

But if we allow custom arguments to packages in the `unit` directory, how do we combine this with the ones that are automatically packaged? Does the automation check whether the attribute already exists and then skip it because it has been manually called? Or do we automatically call everything and the manual call does an additional `.override` onto the automatically called package?

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Author: > It is not because the need for those exceptions doesn't vary over time; only by package.
>
> We can easily tighten up the rules when we actually have the opportunity to do so - when we have a good alternative way to provide default arguments

Those are good points and probably a more general principle to follow, I'll take another think about this.

@piegamesde That's not a problem, we can auto-generate the definitions for all unit directories and just further override the ones with custom arguments. That's how https://github.com/NixOS/nixpkgs/pull/211832 already works.

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Reviewer: `.override` is only possible when the package evaluates with the bad defaults. This makes `callPackage ../unit/...../packages.nix` a more general solution.

I also think it's conceptually simpler that way, but that's subjective because it does rely on the layout. We might abstract the layout as an `attrsOf path` value somewhere, but that's technically unnecessary as well.

> pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`

Author: I reverted this now, non-empty arguments will be allowed in `all-packages.nix`

```

Check the following using CI for each unit directory:
- The only definition for `pkgs.${name}` is the automatically generated one from the unit directory
- `pkgs.${name}` must evaluate to a [derivation](https://nixos.org/manual/nix/stable/glossary.html#gloss-derivation).

> - The `package.nix` file must not transitively refer to files outside its unit directory.

Author: I brought this up in the [last NAT meeting](https://discourse.nixos.org/t/2023-05-02-nixpkgs-architecture-team-meeting-37/27816#notes-1), nobody had any objections to this change, so I'll keep it.

## PR 2: Migration

Automatically migrate to the unit directory standard for all definitions `pkgs.${name}` that can be migrated by
- Only moving files
- Not changing the evaluation result

This will cause merge conflicts with all existing PRs that modify such moved files, however they can trivially be rebased using `git rebase && git push -f`.
However, to have the least amount of conflicts, this migration should be performed soon after a release when ZHF is over and the PR rate slows down.
This also gives a lot of time to fix any potential problems before the next release.

Manual updates may also be done to ensure further non-evaluation validity, such as
- [CODEOWNERS](https://github.com/NixOS/nixpkgs/blob/master/.github/CODEOWNERS)
- Update scripts like [this](https://github.com/NixOS/nixpkgs/blob/cb2d5a2fa9f2fa6dd2a619fc3be3e2de21a6a2f4/pkgs/applications/version-management/cz-cli/generate-dependencies.sh)

> Due to the strict limitations of standard, this PR will not start enforcing it for new packages.

Author: Since [custom arguments are allowed again](https://github.com/NixOS/rfcs/pull/140#discussion_r1191584512), this isn't necessary anymore. I re-added the requirement that new packages must adhere to the standard.

    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)

# Alternatives
[alternatives]: #alternatives

TODO: This needs updating

## A different unit base directory

Context: `pkgs/unit` is the unit base directory

Alternatives:
- Use `unit` (at the Nixpkgs root) instead of `pkgs/unit`.
  - This is future proof in case we want to make the directory structure more general purpose, but this is out of scope

> - Other name proposals were deemed worse: `pkg`, `component`, `part`, `mod`, `comp`

Reviewer 4: **Bikeshedding thread**

"auto" has already been proposed elsewhere, and I'd like to discuss the following alternatives too: "app", "simple".

I like "app" because most of the eligible packages are applications. Of course it might be confusing with libraries, not sure how many packages this affects.

"simple" is meant as a contrast to ecosystem packages like python packages, which make use of additional packaging infrastructure. It is also a hint at most applications in there being "simple" leaf packages

Reviewer: Not to say that we were using this definition, or that you referred to these, but in philosophy, the terms "simple" and "unit" have distinct meanings.

"Simple" refers to something that is not composed of parts. In metaphysics, a simple entity is one that has no internal structure or complexity, and is therefore indivisible. Examples of simple entities include mathematical points, atoms, and souls (in some philosophical traditions).

"Unit," on the other hand, refers to a discrete, indivisible entity that is distinct from other entities of the same type. Units are often thought of as basic building blocks or fundamental constituents of reality. For example, in the philosophy of mathematics, numbers are often thought of as units, and more complex mathematical structures can be built up from these units.

So while both "simple" and "unit" refer to indivisible entities, "simple" refers to something that has no internal structure, while "unit" refers to a discrete entity that can be composed into more complex structures.

In this sense, "unit" is more appropriate for our directories.

Reviewer: I'm not a big fan of this usage of "leaf" by the way. Without context, it _may_ instead refer to the bootstrap, or the derivation sources. I'd prefer "non-dependency" instead: non-dependency package, non-dependency derivation.

Reviewer 4: Did we already discuss `pkgs/default` as a possible name? Nix already uses "default" for file names as a convention, so why not here too? It hopefully evokes the feeling that this is the "default" folder in `pkgs` and therefore packages should put in there by default unless there are specific reasons not to.

## Package variants

Sometimes there's a need to create a variant of a package with different `callPackage` arguments. This can be achieved using `.override` as follows:
```nix
{
  graphviz_nox = graphviz.override { withXorg = false; };
}
```

However this can cause problems with an overlay that tries to make the variant the default as follows:
```nix
self: super: {
  # Oops, infinite recursion!
  graphviz = self.graphviz_nox;
}
```

Because of this, there's the pattern of duplicating the `callPackage` call with the custom arguments as such:
```nix
{
  graphviz_nox = callPackage ../tools/graphics/graphviz { withXorg = false; };
}
```

The semantics of how unit directories are checked by CI do allow the definition of package variants from unit directories:
```nix
{
  graphviz_nox = callPackage ../unit/gr/graphviz/package.nix { withXorg = false; };
}
```

# Drawbacks
[drawbacks]: #drawbacks

- This standard can only be used for top-level packages using `callPackage`, so not for e.g. `python3Packages.requests` or a package defined using `haskellPackages.callPackage`

> - It's not possible anymore to be a [GitHub code owner](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) of category directories.

Reviewer 14: I think that's not a big problem, given that we already have a hacky way to notify people with the `meta.maintainers`.

Reviewer 4: The CODEOWNERS feature is superior to our `meta.maintainers` in a couple of ways, especially since the latter is not always reliable in all circumstances unfortunately. But as much as I am [personally a fan](https://github.com/NixOS/nixpkgs/issues/143441) of using the CODEOWNERS feature more (or an equivalent implemented in GitHub actions), I still think the benefits of this RFC outweigh the cost here.

Reviewer 4: The CODEOWNERS feature is superior to our `meta.maintainers` in a couple of ways, especially since the latter is not always reliable in all circumstances unfortunately. But as much as I am [personally a fan](https://github.com/NixOS/nixpkgs/issues/143441) of using the CODEOWNERS feature more (or an equivalent implemented in GitHub actions), I still think the benefits of this RFC outweigh the cost here.

  - (+) If we remove the other directories at some point, only the `${shard}`'s will be left in `pkgs`
  - (-) This leads to ambiguities between the directories from the standard and the other directories, requiring special handling in the code and CI, leading to complexities.
  - (-) This would be harder to document and explain to people, since one always has to disregard all non-unit directories, with no obvious justification
  - (-) Currently we cannot apply this standard to all definitions in `pkgs`, in particular nested packages like `pythonPackages.*`, non-`callPackage`'d definitions like `copyDesktopItems` and non-derivations like `fetchFromGitHub`.
    Depending on how we want to handle those, it might make more sense to keep `pkgs/unit` or to use `pkgs` directly once all legacy paths are migrated away to another top-level directory, we don't yet know. `pkgs/unit` will be easier to migrate to `pkgs` than the other way around though.

> - Another name than `unit`, proposed were `auto`, `pkg`, `mod`, `component`, `part`, `comp`, `app`, `simple`

Reviewer: ```suggestion
- Another name than `unit`, proposed were `auto`, `pkg`, `mod`, `component`, `part`, `comp`, `app`, `simple`, `shard`
```

Reviewer 15: In light of the current paint-slinging: what about `pkgs/by-name`?

Reviewer 6: @infinisil It would help noting the reasons for each why it's not the preferred choice.

@rhendric `by-name` this is really good, it tells me what it's supposed to mean.

  - (-) `unit` is considered the most appropriate because a "unit" typically refers to a discrete, indivisible entity that is distict from other entities of the same type.
  - (-) In addition we envision that in the future we would extend the unit directory standard to not just include a package definition for each unit, but also other parts such as NixOS modules, library components, tests, etc. In this case `unit` would fit even better and could be described as

>     > A collection of standardized files related to the same software component

Reviewer 4: I don't think I follow the argument against `<nixpkgs>/pkgs/unit`. The way I read the RFC here, expanding the concept to modules in the future does not imply all of them should be put together: I could envisage a solution with `<nixpkgs>/modules/unit` for example.

In general, I quite like the top-level folder structure of nixpkgs and would prefer solutions that don't change it.

>     > A collection of standardized files related to the same software component

Reviewer: > expanding the concept to modules in the future does not imply all of them should be put together

The term `unit` did originate from that idea, in the discussions we've had while drafting this RFC.

Putting related files together is a simplification that helps newcomers most of all. Preserving existing habits would not be a goal, but it would be a consideration.

The `<nixpkgs>/unit` idea is far more radical, so I think we were right to scope it out, but scoping it out does leave a loose end here. Not sure what to do, so I'm inclined to just deliver Simple Package Paths and save "Simple Paths" or whatever for a later RFC.
"Radical": I do think we want to get to the root at some point, figuratively and literally, but perhaps without the connotation.

# Unresolved questions
[unresolved]: #unresolved-questions

# Future work
[future]: #future-work

All of these questions are in scope to be addressed in future discussions in the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html):

- Expose an API to get access to the package functions directly, without calling them
- Add a meta tagging or categorization system to packages as a replacement for the package categories. Maybe `meta.tags` with `search.nixos.org` integration. Maybe https://repology.org/ integration. See also https://github.com/NixOS/rfcs/pull/146.
- Making the filetree more human-friendly by grouping files together by "topic" rather than technical delineations.
  For instance, having a package definition, changelog, package-specific config generator and perhaps even NixOS module in one directory makes work on the package in a broad sense easier.
- This RFC only addresses the top-level attribute namespace, aka packages in `pkgs.<name>`, it doesn't do anything about package sets like `pkgs.python3Packages.<name>`, `pkgs.haskell.packages.ghc942.<name>`, which may or may not also benefit from a similar auto-calling

> - Improve the semantics of `callPackage` and/or apply a better solution, such as a module-like solution

Reviewer 14: None of the solutions discussed so far present an elegant solution, and I don't think one exists. Module solution sounds like it would be quite fitting here (although out of scope of the RFC), so leaving this here for reference:

https://www.youtube.com/watch?v=dTd499Y31ig

I think that would solve the `jami-daemon = jami.jami-daemon` problem by itself, and very possibly `wlroots = wlroots_0_14` if `version` will become an option that has meaning.

Reviewer 16: > Module solution sounds like it would be quite fitting here (although out of scope of the RFC),

Yes, I that's probably out of scope for this PR, but there's relatively fresh working group about exactly this topic. We have a matrix room, weekly meetings on Friday and a Repo to manage things, see https://github.com/nixpkgs-architecture/pkgs-modules

- Potentially establish an updateScript standard to avoid problems like, relates to Flakes too
- What to do with different versions, e.g. `wlroots = wlroots_0_14`? This goes into version resolution, a different problem to fix
- What to do about e.g. `libsForQt5.callPackage`? This goes into overrides, a different problem to fix

> - What about aliases like `jami-daemon = jami.jami-daemon`?

Reviewer 16: There's an answer missing here ;) I believe it might be good enough to keep them in `all-packages.nix`, just as with custom args,  for now, in order to keep the scope small?

Reviewer 17: yeah seems fair

> - What about aliases like `jami-daemon = jami.jami-daemon`?

Reviewer: > To keep the scope small, that would be an idea for another follow-up RFC.

I agree.

I believe such an idea has been considered, but was rejected for scope reasons, and a potential performance issue.

As these files would each be able to affect the top level `attrNames`, they would all have to be parsed before constructing "pkgs". So far we've kept the proposal from having to scan the _contents of_ all unit directories when evaluating a limited set of attribute values. Actually parsing files in many of them would change that. Most practical usages do not involve evaluating all of Nixpkgs, so we should optimize to keep "individual" package evaluations fast.
The actual performance impact has not been measured though, nor have we assessed the changes on the human side.

  - (-) In addition we envision that in the future we would extend the unit directory standard to not just include a package definition for each unit, but also other parts such as NixOS modules, library components, tests, etc. In this case `unit` would fit even better and could be described as
    > A collection of standardized files related to the same software component

## Alternate shard structure

Context: The structure is `pkgs/unit/${shard}/${name}` with `shard` being the lowercased two-letter prefix of `name`.

Alternatives:
- A flat directory, where `pkgs.hello` would be in `pkgs/unit/hello`.
  - (+) Simpler for the user and code.
  - (-) The GitHub web interface only renders the first 1 000 entries when browsing directories, which would make most packages inaccessible in this way.
    - (+) This feature is not used often.
      - (-) [A poll](https://discourse.nixos.org/t/poll-how-often-do-you-navigate-through-directories-on-github-to-get-to-a-nixpkgs-package-file/21641) showed that about 41% of people rely on this feature every week.
  - (-) Bad because it makes `git` and file listings slower.
- Use three-letter or four-letter prefixes.

>   - (-) Also leads to directories containing more than 1 000 entries, see above.

Reviewer 15: Anyone considered using, say, the first and last character of the name instead of a prefix? I tried that and the largest shard had 202 entries (though I was lazy and just checked all the attr names of the top-level object, instead of doing whatever is the right thing for listing all packages and only packages). Still only two characters and one level, still fully determined by the name alone.

Author: Interesting idea! By adjusting the [gist to calculate the distribution of shards](https://gist.github.com/infinisil/95c7013db62e9f23ab2bc92165a37221) to this:

<details>

```
{
  system ? builtins.currentSystem,
  nixpkgs ? fetchTarball {
    url = "https://github.com/NixOS/nixpkgs/tarball/6948ef4deff7a72ebe5242244bd3730e8542b925";
    sha256 = "1fw8vlbcxr3kw96rrjlw94zhbjs3a32dcnvav19cvniwlbhf2jsh";
  },
}:
let
  pkgs = import nixpkgs {
    inherit system;
    config.allowAliases = false;
    overlays = [];
  };
  inherit (pkgs) lib;

  set = pkgs;

  names = lib.attrNames set;
  attributeCount = lib.length names;

  grouped = lib.groupBy (name:
    lib.toLower (lib.substring 0 1 name + lib.substring (lib.stringLength name - 1) 1 name)
  ) names;

  distribution = lib.sort (a: b: a.count < b.count)
    (lib.mapAttrsToList (name: values: {
      inherit name;
      count = lib.length values;
    }) grouped);

  shardCount = lib.length distribution;

  top20 = lib.drop (shardCount - 20) distribution;

  result =
    builtins.trace "Attribute count: ${toString attributeCount}"
    builtins.trace "Shard count: ${toString shardCount}"
    builtins.trace "Top 20"
    (lib.foldl' (acc: el:
      builtins.trace "${el.name}: ${toString el.count}" acc
    ) null top20);

in {
  inherit result;
}
```
</details>

I'm getting this distribution:

```
$ nix-instantiate shards.nix --eval -A result
trace: Attribute count: 17305
trace: Shard count: 895
trace: Top 20
trace: lr: 104
trace: pt: 104
trace: gt: 106
trace: as: 108
trace: mr: 110
trace: se: 110
trace: pe: 113
trace: ms: 114
trace: ce: 120
trace: lt: 129
trace: ge: 134
trace: st: 134
trace: ls: 135
trace: cs: 136
trace: sr: 137
trace: le: 138
trace: gs: 155
trace: ss: 160
trace: pr: 171
trace: ps: 183
```

For comparison, here's the one with the 2-letter prefix:

<details>

```
trace: Shard count: 726
trace: Top 20
trace: gr: 121
trace: do: 133
trace: mi: 133
trace: gn: 137
trace: ne: 139
trace: po: 143
trace: st: 144
trace: mo: 149
trace: cl: 161
trace: re: 166
trace: gi: 172
trace: pa: 184
trace: te: 184
trace: go: 210
trace: pr: 217
trace: ca: 243
trace: ma: 244
trace: co: 252
trace: op: 260
trace: li: 1092
```
</details>

And simulating a larger package count using
```nix
pkgs // pkgs.haskellPackages // pkgs.python3Packages // pkgs.perlPackages // pkgs.emacsPackages;
```

```
trace: Attribute count: 47078
trace: Shard count: 951
trace: Top 20
trace: sr: 348
trace: de: 351
trace: ae: 357
trace: st: 361
trace: ts: 366
trace: me: 367
trace: ms: 367
trace: he: 368
trace: gs: 369
trace: ge: 390
trace: hs: 401
trace: pr: 403
trace: pt: 412
trace: as: 423
trace: cs: 433
trace: ce: 489
trace: se: 551
trace: ss: 571
trace: ps: 574
trace: pe: 583
null
```

And the 2-letter prefix for comparison:
<details>

```
trace: Attribute count: 47078
trace: Shard count: 804
trace: Top 20
trace: da: 405
trace: he: 413
trace: op: 436
trace: fl: 441
trace: po: 474
trace: se: 479
trace: di: 487
trace: go: 572
trace: st: 573
trace: pr: 587
trace: pa: 603
trace: mo: 606
trace: ma: 610
trace: ca: 628
trace: te: 685
trace: ha: 733
trace: re: 738
trace: co: 1031
trace: py: 1194
trace: li: 1423
```
</details>

It's really nice how the distribution is much more equal, keeping the number of directories in shards low even with many more packages!

Though I also see a number of disadvantages:
- It's not as obvious what the sharding scheme is. 2-letter prefix is slightly more obvious from the paths themselves (`pkgs/unit/he/hello`), and much more obvious when seeing all packages in the same directory having the same prefix (`pkgs/unit/he` only contains entries starting with `he`)
- Similarly-named packages won't be close to each other anymore, `hello` will be in `ho`, but `hello-unfree` in `he`.
- The directory listing for `pkgs/unit` won't be attribute-alphabetical
- Changing the suffix of attributes (or creating new ones with a different suffix) requires moving them to a different shard.

Reviewer 4: I find the approach really interesting, but for me the major downside is the added cognitive overhead when navigating — even if I know the scheme. Try to pick a random package name (preferably without reading its name on screen so that you have to do the calculation in your head) and type its first two letters vs its first and last.

But this got me thinking: seeing both lists @infinisil generated, to me they look pretty similar except for `li` because of `lib`. So how about we hard-code special-case some exceptions, so that packages that start with `lib` get a five letter shard? (Alternatively, strip lib from the name before sharding so `libfoo` would land in `fo` instead of `li`)

Author: I don't think it's a big problem that the `li` shard has over 1000 entries. Yeah GitHub doesn't display all of them in the directory view, but that's very minor (only about 100 packages aren't reachable directly from the file browser!). In comparison, any exception would add a non-trivial mental overhead and potential confusion. So I don't think we should do that.

Reviewer 4: It's always a tradeoff, and I have no strong opinions here. We should document these idea of the Alternatives section though.

Reviewer 17: i think doing an exception for lib is trivial enough that the benefits are greater than the added mental-overhead

Reviewer 4: I'd like to explore the idea of a five-letter shard for "lib" some more, and would like to see the result of the resulting distribution. My main concern here is that in theory it may result in more than 1000 shards in the `unit` directory, although I doubt we'll get close to that in practice anytime soon.

Reviewer 17: if we do `lib.toLower (lib.substring 1 2 name)`, we get `ib` as largest shard with 935, which would fit into the github limit, but that would be ridiculous and not less confusing than using the first and last character
although the grouping of e.g. `hello` next to `hello_unfree` would work

Author: I just noticed that already today we have three directories in Nixpkgs with more than 1000 entries!
- [`pkgs/development/libraries`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/development/libraries) with 1995 entries (my birth year! :birthday:), so probably already about 995 libraries are currently unavailable directly from GitHub's file browser!
- [`pkgs/development/python-modules`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/development/python-modules) with 6145 entries! This is currently an effectively flat listing of all python packages, so this would be fixed if we used the same sharding in the future.
- [`pkgs/tools/misc`](https://github.com/NixOS/nixpkgs/tree/7d9999f2c178882a4656283c12a7501243626bf7/pkgs/tools/misc) with 1013 entries, an additional 13 entries not directly accessible

And I don't think I've ever seen anybody complain about this, so I don't think we should be worried about this proposal ending up with only ~100 non-directly accessible packages.

Reviewer 17: Although i don't use GitHub to browse the Nixpkgs source code anymore (i navigate to package sources by clicking on `Source` on search.nixos.org), i found the GitHub limitations quite puzzling and frustrating when i first started out with Nix.

Reviewer 15: > Alternatively, strip lib from the name before sharding so `libfoo` would land in `fo` instead of `li`.

I liked this thought; it shouldn't blow up the number of shards, and if putting related packages in the same shard is a goal then any `foo`/`libfoo` pairs we have will end up together.

I wouldn't object to any of the ideas proposed, but like @Quantenzitrone I do find it occasionally frustrating not to be able to make maximum use of GitHub's affordances.

Reviewer 17: A downside of making an exception for lib* packages, are packages, that start with lib but are not a library like e.g. `librarian-puppet-go`.

Reviewer 4: It also opens up the possibility for collisions, e.g. `libfoo` vs `foo`, and not that unlikely ones.

Reviewer 17: no, the `name` will still be `libfoo`, so no collision with `foo` happens, otherwise the autocaller wouldn't that libfoo is a library
e.g.
```
fo
|- foo
|- foobar
'- libfoo
```

Reviewer 18: The flipside to the GitHub limitations being annoying is that, if GitHub being annoying pushes somebody from navigating on GitHub to navigating a local clone, they're one step closer to contributing!

Reviewer 17: if that is the goal, we should use the flat directory structure proposed in the alternatives

Reviewer 17: > I'd like to explore the idea of a five-letter shard for "lib" some more, and would like to see the result of the resulting distribution. My main concern here is that in theory it may result in more than 1000 shards in the `unit` directory, although I doubt we'll get close to that in practice anytime soon.

how about a 4 letter shard for `lib*` packages tho?
there are only 909 packages that start with lib, so it doesn't really make sense to split these into 323 different shards, but when using 4 letter shards for `lib*` packages we only have 29 shards for the lib packages with `libs` being the biggest shard with 70 packages
this results in only 755 total shards and has the benefit of it being easier to understand due to the shard being the start of the package attribute name (just a longer substring for lib*)

Author: Another argument for not adding such a special case: We can expect to eventually want to use this standard for other package sets too, such as Python packages. Keeping the same special case as for top-level packages doesn't make sense then. One might argue that we should be able to add arbitrary special cases depending on the package set, but that does introduce another variable to the game, increasing complexity again.

From the arguments brought up so far I'm fully convinced that we shouldn't do this, the benefits are not high enough to justify the cost.

Reviewer 17: yeah for python we would need a special case for the py shard, as it currently has about 1070 packages

Author: Thinking about this some more, there's a good case to be made that all these library packages should really be in their own package set, since they're not generally something users would directly install (which is somewhat of a convention with top-level packages). So we might want to have `lib.png` instead of `libpng`.

Reviewer 18: I don't agree with this.  The convention is not about whether something is directly installable, but whether it makes sense for it to be user accessible, and whether it belongs in a separate namespace to avoid collisions (e.g. `pythonPackages`).  Directly using (albeit not installing) a library package is extremely common: `nix-shell -p libpng` when compiling a small program.  Such a change would make Nixpkgs pointlessly different to every other package repository.

Reviewer 18: I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

Author: The structure of Nixpkgs doesn't have to be exposed directly to the user though. I could imagine a future in which we expose packages both stripped (for a clean Nixpkgs tree with unit directories) and unstripped (for upstream names). Anyways, this is all just an "could be", not a "have to". I'll make sure to add all the arguments mentioned here to the RFC with that in mind.

Reviewer 17: > I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

actually for some package sets we do, look at fishPlugins for example

Reviewer 17: > Thinking about this some more, there's a good case to be made that all these library packages should really be in their own package set, since they're not generally something users would directly install (which is somewhat of a convention with top-level packages). So we might want to have `lib.png` instead of `libpng`.

How exactly would a library be defined? With package sets like pythonPackages or vimPlugins, this is mostly unambiguous, but for libraries, anything can have a `lib` output (right?).
Furthermore, not all libraries start with lib.

Reviewer 17: i honestly would just leave the unit structure as it is right now, having one shard with over 1000 entries is acceptable and definitely better than the current package structure, so let's leave possible shard exceptions to later rfc's and get this merged

Reviewer 19: > I'll also note that we don't strip prefixes even for packages in namespaces — look at all the package names starting with "python-" in python-packages.nix.

This is generally done on purpose to follow the upstream/PyPI names.

Reviewer 4: Have we discussed single-letter prefix already? Someone argued that the 1000 file folder limit of GitHub is pretty arbitrary as restriction, and while I do think we should try to bound the folder size, I'd like to know how bad it gets with single-letter prefixes.

Reviewer 17: ```
trace: Attribute count: 17305
trace: Shard count: 27
trace: Top 20
trace: _: 31
trace: y: 134
trace: q: 210
trace: z: 213
trace: j: 270
trace: v: 311
trace: u: 340
trace: x: 354
trace: k: 399
trace: i: 426
trace: w: 430
trace: h: 440
trace: e: 479
trace: o: 529
trace: n: 634
trace: b: 647
trace: r: 662
trace: f: 691
trace: d: 708
trace: t: 801
trace: a: 854
trace: m: 1039
trace: p: 1217
trace: c: 1246
trace: g: 1268
trace: s: 1424
trace: l: 1548
```

  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
    - It also slows down or even locks up editors due to the file size
  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
These PR's should be done after a release to maximize the testing period and minimize merge conflicts.

## PR 1: The directory structure

This part establishes the new directory structure in Nixpkgs.
This directory structure is internal to Nixpkgs and not exposed as public interface.
This directory structure must be documented in the Nixpkgs manual.
This PR will be backported to the stable release in order to ensure that backports of new packages work.

### File structure

> Create the initially-empty _base directory_ `pkgs/by-name` in Nixpkgs.

Reviewer: Nitpick 1: git won't let you do that, so move `hello` in.

Nitpick 2: we barely use the term _base directory_ in this document and we should not refer to it outside this document, because that would be ambiguous. We may as well use `pkgs/by-name` in the remaining places.

```suggestion
Create the initially-empty `pkgs/by-name` directory in Nixpkgs, and migrate the `hello` package into it.
```

Check the following using CI:

> - The base directory must only contain subdirectories of the form `${shard}/${name}`, called _package directories_.

Reviewer: ```suggestion
- `pkgs/by-name` must only contain subdirectories of the form `${shard}/${name}`, called _package directories_.
```

Continuing the removal of the hardly used term _base directory_.

- The `name`'s of package directories must be unique when lowercased
- `name` is a string only consisting of the ASCII characters `a-z`, `A-Z`, `0-9`, `-` or `_`.
- `shard` is the lowercased first two letters of `name`, expressed in Nix: `shard = toLower (substring 0 2 name)`.
- Each package directory must contain a `package.nix` file and may contain arbitrary other files.

### Semantics

> Introduce code to automatically define `pkgs.${name}` for each package directory as a value equivalent to

Reviewer 20: (obviously a bit late to the party, sorry about that)

wouldn't this code generate a ton of `getdents` syscalls? (and degrades performance of evaluation over time as packages are added)

Reviewer 4: Please have a look at the performance discussion that happened on the implementation PRs.

Reviewer 20: I checked the two other PRs, but didn't check this one. Thanks!

https://github.com/NixOS/nixpkgs/pull/237439#issuecomment-1640686296 (and following comments)

- The existing categorization of packages gets lost. Counter-arguments:
  - It was never that useful to begin with.
    - The categorization was always incomplete, because packages defined in the language package sets often don't get their own categorized file path.
    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
- This breaks `builtins.unsafeGetAttrPos "hello" pkgs`. Counter-arguments:
  - We have to draw a line as to what constitutes the public interface of Nixpkgs. We have decided that making attribute position information part of that is not productive. For context, this information is already accepted to be unreliable at the language level, noting the `unsafe` part of the name.
  - Support for this could be added to Nix (make `builtins.readDir` propagate file as a position)

# Alternatives
[alternatives]: #alternatives

> ## A different base directory
> 
> Context: `pkgs/by-name` is the base directory

Reviewer: ```suggestion
## An alternative to the `pkgs/by-name` location

Context: this directory contains the shards, which contain the package directories. We could move the shards to a different location.
```

Continuing the removal of the hardly used term _base directory_.

  - (-) Requires special-casing single-letter attribute names
    - (+) There's currently only 6 such cases, which could be handled on a one-off basis
  - (-) Makes auto-completion worse, having to tab-complete once more
  - (-) Makes it harder to create shards: if a shard doesn't exist yet, it has to be created with either one or two `mkdir`'s, or a `mkdir -p`
- Use a dynamic structure where directories are rebalanced when they have too many entries.
  E.g. `pkgs.foobar` could be in `pkgs/by-name/f/foobar` initially.
  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/by-name/fo/foobar`.
  - (-) The structure depends not only on the name of the package then, making it harder to find packages again and figure out where they should go
  - (-) Complex to implement

## Alternate `package.nix` filename

Context: The only file that has to exist in package directories is `package.nix`, it must contain a function suitable for `callPackage`.

Alternatives:
- `default.nix`
  - (+) `default.nix` is already a convention most people are used to.

>   - (-) We don't benefit from the usual `default.nix` benefits:

Reviewer 11: We would benefit that people are used to placing default.nix files in directories. It is the defacto filename which IMO should have just continued to be used.

    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
      - (+) But there's still some support for `all-packages.nix` for custom arguments, which requires people to type out the name
        - (-) This is hopefully only temporary, in the future we should fully get rid of `all-packages.nix`
    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/by-name/hell/hello` equally broken regardless of file name.
  - (-) Not using `default.nix` frees up `default.nix` for an expression that is actually buildable, e.g. `(import ../.. {}).hello`, although we don't yet have a use case for this that isn't covered by `nix-build ../.. -A <attrname>`.

>   - (-) Using `default.nix` would tempt users to invoke `nix-build .`, which wouldn't work and making package functions auto-callable is a known anti-pattern.

Reviewer 11: I doubt people are actually doing that and this is something to worry about. Everyone I have observed and seen goes though all-packages.nix except when doing out of tree things which doesn't apply here anyway.

Reviewer 21: > I doubt people are actually doing that and this is something to worry about. Everyone I have observed and seen goes though all-packages.nix except when doing out of tree things which doesn't apply here anyway.

The additional benefit of `package.nix` is that if one day nixos module files are put in the same dir as the package

Reviewer 16: This RFC has been accepted & merged after extensive discussion several months ago, so I might be a bit late to add this input here. As I understand the project, the correct way would be to create a new RFC to suggest your desired change if you care strongly enough about it.

Author: Indeed, and you already brought this up before [here](https://github.com/NixOS/rfcs/pull/140#discussion_r1090915765)

  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
    - It also slows down or even locks up editors due to the file size
  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages

# Detailed design
[design]: #detailed-design

This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
These PR's should be done after a release to maximize the testing period and minimize merge conflicts.

## PR 1: The unit directory standard

This part establishes the new _unit directory standard_ in Nixpkgs.
This standard is internal to Nixpkgs and not exposed as public interface.
This standard must be documented in the Nixpkgs manual.
This PR will be backported to the stable release in order to ensure that backports of new packages work.

### The name "unit"

> The term "unit" is chosen partly because it's not associated with any pre-existing assumptions about what it means, which should cause people unfamiliar with this standard to read the documentation.

Reviewer 2: This is false, we already have a different meaning of «unit»: https://nixos.org/manual/nixos/stable/index.html#sec-unit-handling

Reviewer: Systemd has one, and so do the sciences, and mathematics has a couple kinds of "units". Organizations have units... They're all different.

It's a very broad term, so I wouldn't say that using it for something else would be unexpected. That said, we should weaken the language here a bit.

```suggestion
The term "unit" is chosen partly because it's generally not associated with any pre-existing assumptions about what it means, which should cause people unfamiliar with this standard to read the documentation.
```

Check the following using CI for each unit directory:
- `pkgs.${name}` is defined as above, either automatically or with some `args` in `pkgs/top-level/all-packages.nix`.
- `pkgs.${name}` is a [derivation](https://nixos.org/manual/nix/stable/glossary.html#gloss-derivation).

> - <a id="req-ref"> The `package.nix` file must not transitively refer to files outside its unit directory.

Reviewer 22: probably doesn't matter in markdown but:

![image](https://github.com/NixOS/rfcs/assets/9866621/520f099a-c398-48e8-852c-e045615ae45c)

```suggestion
- <a id="req-ref"></a> The `package.nix` file must not transitively refer to files outside its unit directory.
```

(if this is meant to be an invisible anchor)

Author: Thanks, fixed!
